File: 1-two-sum/two-sum.php
class Solution {

    /**
     * @param Integer[] $nums
     * @param Integer $target
     * @return Integer[]
     */
    function twoSum($nums, $target) {
        $numMap = [];  // value => index
        
        foreach ($nums as $index => $num) {
            $complement = $target - $num;
            
            // Check if complement exists in our map
            if (array_key_exists($complement, $numMap)) {
                return [$numMap[$complement], $index];
            }
            
            // Store current number with its index
            $numMap[$num] = $index;
        }
        
        return [];  // No solution found (though problem guarantees one exists)
    }
}

File: 1-two-sum/two-sum.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[]}
def two_sum(nums, target)
    num_to_index = {}
    
    nums.each_with_index do |num, index|
        complement = target - num
        
        if num_to_index.key?(complement)
            return [num_to_index[complement], index]
        end
        
        num_to_index[num] = index
    end
    
    # Since the problem guarantees exactly one solution, we'll always return above
    []
end